server/src/main/java/org/elasticsearch/cluster/version/CompatibilityVersions.java

```diff
@@ -13,49 +13,107 @@
- * This wrapper lets us wrap these versions one level below {@link org.elasticsearch.cluster.ClusterState}.
- * It's similar to {@link org.elasticsearch.cluster.node.VersionInformation}, but this class is meant to
- * be constructed during node startup and hold values from plugins as well.
+ * This wrapper lets us wrap these versions one level below {@link org.elasticsearch.cluster.ClusterState}. It's similar to
+ * {@link org.elasticsearch.cluster.node.VersionInformation}, but this class is meant to be constructed during node startup and hold values
+ * from plugins as well.
  *
- * @param transportVersion A transport version, usually a minimum compatible one for a node.
+ * @param transportVersion           A transport version, usually a minimum compatible one for a node.
+ * @param systemIndexMappingsVersion A map of system index names to versions for their mappings.
  */
-public record CompatibilityVersions(TransportVersion transportVersion) implements Writeable, ToXContentFragment {
+public record CompatibilityVersions(
+    TransportVersion transportVersion,
+    Map<String, org.elasticsearch.indices.SystemIndexDescriptor.MappingsVersion> systemIndexMappingsVersion
+) implements Writeable, ToXContentFragment {
 
     /**
      * Constructs a VersionWrapper collecting all the minimum versions from the values of the map.
      *
      * @param compatibilityVersions A map of strings (typically node identifiers) and versions wrappers
      * @return Minimum versions for the cluster
      */
-    public static CompatibilityVersions minimumVersions(Map<String, CompatibilityVersions> compatibilityVersions) {
-        return new CompatibilityVersions(
-            compatibilityVersions.values()
-                .stream()
-                .map(CompatibilityVersions::transportVersion)
-                .min(Comparator.naturalOrder())
-                // In practice transportVersions is always nonempty (except in tests) but use a conservative default anyway:
-                .orElse(TransportVersions.MINIMUM_COMPATIBLE)
-        );
+    public static CompatibilityVersions minimumVersions(Collection<CompatibilityVersions> compatibilityVersions) {
+        TransportVersion minimumTransport = compatibilityVersions.stream()
+            .map(CompatibilityVersions::transportVersion)
+            .min(Comparator.naturalOrder())
+            // In practice transportVersions is always nonempty (except in tests) but use a conservative default anyway:
+            .orElse(TransportVersions.MINIMUM_COMPATIBLE);
+
+        Map<String, SystemIndexDescriptor.MappingsVersion> minimumMappingsVersions = new HashMap<>();
+        compatibilityVersions.stream()
+            .flatMap(mv -> mv.systemIndexMappingsVersion().entrySet().stream())
+            .forEach(
+                entry -> minimumMappingsVersions.merge(entry.getKey(), entry.getValue(), (v1, v2) -> v1.version() < v2.version() ? v1 : v2)
```

Comment:
You could make `MappingsVersion` implement `Comparable` using the version number so an ordering is part of the class definition

Reply:
That's a good idea. It'll make comparisons in downstream PRs look a lot cleaner.